packages/shared/src/components/HorizontalScroll/useHorizontalScrollHeader.tsx

```diff
@@ -0,0 +1,127 @@
+import React, {
+  MouseEventHandler,
+  ReactElement,
+  RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import { useScrollManagement } from './useScrollManagement';
+import { Button, ButtonVariant } from '../buttons/Button';
+import { ArrowIcon } from '../icons';
+
+interface HorizontalScrollHeaderReturn {
+  Header: React.ComponentType<React.HTMLAttributes<HTMLDivElement>>;
+  ref: React.RefObject<HTMLDivElement>;
+}
+
+export interface HorizontalScrollHeaderProps {
+  onScroll?: (ref: RefObject<HTMLElement>) => void;
+  onClickSeeAll?: MouseEventHandler;
+  title: string | ReactElement;
+}
+
+export const useHorizontalScrollHeader = ({
+  onScroll,
+  onClickSeeAll,
+  title,
+}: HorizontalScrollHeaderProps): HorizontalScrollHeaderReturn => {
+  const ref = useRef<HTMLDivElement>(null);
+  const [scrollableElementWidth, setScrollableElementWidth] =
+    useState<number>(0);
+  const [numCards, setNumCards] = useState<number>(0);
+
+  // Calculate the width of elements and the number of visible cards
+  const calculateVisibleCards = useCallback(() => {
```

Comment:
This seems like a highly expensive calculation right?
do we really need to do it as a "seperate one" meaning, maybe we can just use the computedStyle of the main container? (as you already know how many children you render)